checksum.c

```diff
@@ -134,18 +134,20 @@ create_osi_cksum (const uint8_t *pptr, int checksum_offset, int length)
 
     mul = (length - checksum_offset)*(c0);
 
-    x = mul - c0 - c1;
-    y = c1 - mul - 1;
-
-    if ( y >= 0 ) y++;
-    if ( x < 0 ) x--;
+    /*
+     * Casting c0 and c1 here is guaranteed to be safe, because we know
+     * they have values between 0 and 254 inclusive.  These casts are
+     * done to ensure that all of the arithmetic operations are
+     * well-defined (i.e., not mixing signed and unsigned integers).
+     */
+    x = mul - (int)c0 - (int)c1;
+    y = (int)c1 - mul;
 
     x %= 255;
     y %= 255;
 
-
-    if (x == 0) x = 255;
-    if (y == 0) y = 255;
+    if (x <= 0) x += 255;
+    if (y <= 0) y += 255;
 
     y &= 0x00FF;
     checksum = ((x << 8) | y);
```

Comment:
@fxlb asked:
> There is a warning with Visual Studio 17 2022:
checksum.c(153,26): warning C4244: '=': conversion from 'int' to 'uint16_t', possible loss of data.
https://ci.appveyor.com/project/guyharris/tcpdump/builds/51159591/job/or4dequ9f8xe2n4v
Cast needed?

This line hasn't changed. Before my changes, x and y were ints, and checksum was uint16_t.

I wouldn't argue with the introduction of a (uint16_t) cast here. I don't think my change added this warning.